// @nx-extensions/helm/src/generators/chart/generator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.chartGenerator = chartGenerator;
const tslib_1 = require("tslib");
const node_path_1 = tslib_1.__importDefault(require("node:path"));
const devkit_1 = require("@nx/devkit");
function chartGenerator(tree, options) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        var _a;
        const project = (0, devkit_1.readProjectConfiguration)(tree, options.project);
        if ((_a = project.targets) === null || _a === void 0 ? void 0 : _a.helm) {
            throw new Error(`Project ${options.project} already has a helm target. Please remove it before running this command.`);
        }
        (0, devkit_1.updateProjectConfiguration)(tree, options.project, Object.assign(Object.assign({}, project), { targets: Object.assign(Object.assign({}, project.targets), { helm: {
                    executor: '@nx-extensions/helm:package',
                    outputs: ['{options.outputFolder}'],
                    options: {
                        chartFolder: `${project.root}/${options.chartFolder}`,
                        outputFolder: '{workspaceRoot}/dist/charts/{projectRoot}',
                        push: false,
                        remote: 'oci://localhost:5000/helm-charts',
                        dependencies: {
                            update: true,
                            build: true,
                            repositories: [],
                        },
                    },
                } }) }));
        (0, devkit_1.generateFiles)(tree, 
        // eslint-disable-next-line unicorn/prefer-module
        node_path_1.default.join(__dirname, 'files', 'chart'), node_path_1.default.join(project.root, options.chartFolder), options);
        if (options.format) {
            yield (0, devkit_1.formatFiles)(tree);
        }
        // TODO: can we do this programmatically?
        devkit_1.logger.warn(`Add ${project.root}/${options.chartFolder} to .prettierignore to avoid Helm chart syntax errors.`);
    });
}
exports.default = chartGenerator;
