// @nx-tools/nx-container/dist/plugins/nodes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createNodes = exports.createNodesV2 = exports.createDependencies = exports.PLUGIN_NAME = void 0;
const devkit_1 = require("@nx/devkit");
const calculate_hash_for_create_nodes_1 = require("@nx/devkit/src/utils/calculate-hash-for-create-nodes");
const js_1 = require("@nx/js");
const node_fs_1 = require("node:fs");
const cache_directory_1 = require("nx/src/utils/cache-directory");
const path_1 = require("path");
const constants_1 = require("../generators/configuration/constants");
exports.PLUGIN_NAME = '@nx-tools/nx-container';
const cachePath = (0, path_1.join)(cache_directory_1.workspaceDataDirectory, 'container.hash');
const targetsCache = readTargetsCache();
function readTargetsCache() {
    return (0, node_fs_1.existsSync)(cachePath) ? (0, devkit_1.readJsonFile)(cachePath) : {};
}
function writeTargetsToCache() {
    (0, devkit_1.writeJsonFile)(cachePath, targetsCache);
}
const createDependencies = () => {
    writeTargetsToCache();
    return [];
};
exports.createDependencies = createDependencies;
exports.createNodesV2 = [
    '**/Dockerfile',
    async (configFiles, options, context) => {
        return await (0, devkit_1.createNodesFromFiles)((configFile, options, context) => createNodesInternal(configFile, options, context), configFiles, options, context);
    },
];
exports.createNodes = exports.createNodesV2;
async function createNodesInternal(configFilePath, options, context) {
    const normalized = normalizeOptions(options);
    const projectRoot = (0, path_1.dirname)(configFilePath);
    // Do not create a project if package.json and project.json isn't there.
    const isProject = (0, node_fs_1.existsSync)((0, path_1.join)(projectRoot, 'project.json')) || (0, node_fs_1.existsSync)((0, path_1.join)(projectRoot, 'package.json'));
    if (!isProject) {
        return {};
    }
    const hash = await (0, calculate_hash_for_create_nodes_1.calculateHashForCreateNodes)(projectRoot, normalized, context, [
        (0, js_1.getLockFileName)((0, devkit_1.detectPackageManager)(context.workspaceRoot)),
    ]);
    const projectName = buildProjectName(projectRoot, context.workspaceRoot, normalized);
    targetsCache[hash] ??= buildTargets(projectRoot, normalized, projectName);
    return {
        projects: {
            [projectRoot]: {
                targets: targetsCache[hash],
            },
        },
    };
}
function buildTargets(projectRoot, options, projectName) {
    const targets = {
        [options.buildTargetName]: {
            executor: '@nx-tools/nx-container:build',
            dependsOn: ['build'],
            options: {
                engine: options.defaultEngine,
                tags: [`${projectName}:dev`],
                load: true,
            },
            configurations: {
                ci: {
                    load: false,
                    push: true,
                    metadata: {
                        images: [projectName],
                        tags: [
                            'type=schedule',
                            'type=ref,event=branch',
                            'type=ref,event=tag',
                            'type=ref,event=pr',
                            'type=sha,prefix=sha-',
                        ],
                    },
                },
            },
        },
    };
    return targets;
}
function buildProjectName(projectRoot, workspaceRoot, options) {
    const packageJsonPath = (0, path_1.join)(projectRoot, 'package.json');
    const projectJsonPath = (0, path_1.join)(projectRoot, 'project.json');
    let name = '';
    const registry = options.defaultRegistry?.length ? `${options.defaultRegistry}/` : '';
    if ((0, node_fs_1.existsSync)(projectJsonPath)) {
        const projectJson = (0, devkit_1.parseJson)((0, node_fs_1.readFileSync)(projectJsonPath, 'utf-8'));
        name = `${registry}${projectJson.name}`;
    }
    else if ((0, node_fs_1.existsSync)(packageJsonPath)) {
        const packageJson = (0, devkit_1.parseJson)((0, node_fs_1.readFileSync)(packageJsonPath, 'utf-8'));
        name = `${registry}${packageJson.name}`;
    }
    return name.replace(/@/gi, '');
}
function normalizeOptions(options) {
    return {
        buildTargetName: options?.buildTargetName ?? 'container',
        defaultEngine: options?.defaultEngine ?? constants_1.DEFAULT_ENGINE,
        defaultRegistry: options?.defaultRegistry ?? constants_1.DEFAULT_REGISTRY,
    };
}
