// @nx-tools/nx-container/dist/executors/build/engines/podman/podman.engine.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Podman = void 0;
const tslib_1 = require("tslib");
const core_1 = require("@nx-tools/core");
const handlebars = tslib_1.__importStar(require("handlebars"));
const engine_adapter_1 = require("../engine-adapter");
const podman = tslib_1.__importStar(require("./podman"));
class Podman extends engine_adapter_1.EngineAdapter {
    podmanVersion = '';
    getPodmanVersion() {
        return this.podmanVersion;
    }
    getCommand(args) {
        return {
            command: 'podman',
            args,
        };
    }
    async initialize(inputs, _ctx) {
        if (!(await podman.isAvailable())) {
            throw new Error(`Podman is required. See https://github.com/gperdomor/nx-tools to set up nx-container executor with podman.`);
        }
        if (!inputs.quiet) {
            await core_1.logger.group(`Podman info`, async () => {
                await (0, core_1.exec)('podman', ['version'], {
                    failOnStdErr: false,
                });
                await (0, core_1.exec)('podman', ['info'], {
                    failOnStdErr: false,
                });
            });
        }
        this.podmanVersion = await podman.getVersion();
        if (!inputs.quiet) {
            await core_1.logger.group(`Podman version`, async () => {
                await (0, core_1.exec)('podman', ['version'], {
                    failOnStdErr: false,
                });
            });
        }
    }
    async finalize(inputs, _ctx) {
        if (inputs.push) {
            await (0, core_1.asyncForEach)(inputs.tags, async (tag) => {
                const cmd = this.getCommand(['push', tag]);
                await (0, core_1.getExecOutput)(cmd.command, cmd.args, {
                    ignoreReturnCode: true,
                }).then((res) => {
                    if (res.stderr.length > 0 && res.exitCode != 0) {
                        throw new Error(`podman failed with: ${res.stderr.match(/(.*)\s*$/)?.[0]?.trim() ?? 'unknown error'}`);
                    }
                });
            });
        }
    }
    async getImageID() {
        return podman.getImageID();
    }
    async getMetadata() {
        return podman.getMetadata();
    }
    async getDigest(metadata) {
        return podman.getDigest(metadata);
    }
    async getArgs(inputs, defaultContext) {
        const context = handlebars.compile(inputs.context)({ defaultContext });
        // prettier-ignore
        return [
            ...await this.getBuildArgs(inputs, defaultContext, context, this.getPodmanVersion()),
            ...await this.getCommonArgs(inputs, this.getPodmanVersion()),
            context
        ];
    }
    async getBuildArgs(inputs, defaultContext, context, podmanVersion) {
        const args = ['build'];
        await (0, core_1.asyncForEach)(inputs.addHosts, async (addHost) => {
            args.push('--add-host', addHost);
        });
        await (0, core_1.asyncForEach)(inputs.buildArgs, async (buildArg) => {
            args.push('--build-arg', buildArg);
        });
        if (podman.satisfies(podmanVersion, '>=4.2.0')) {
            await (0, core_1.asyncForEach)(inputs.buildContexts, async (buildContext) => {
                args.push('--build-context', buildContext);
            });
        }
        await (0, core_1.asyncForEach)(inputs.cacheFrom, async (cacheFrom) => {
            args.push('--cache-from', cacheFrom);
        });
        await (0, core_1.asyncForEach)(inputs.cacheTo, async (cacheTo) => {
            args.push('--cache-to', cacheTo);
        });
        if (inputs.cgroupParent) {
            args.push('--cgroup-parent', inputs.cgroupParent);
        }
        if (inputs.file) {
            args.push('--file', inputs.file);
        }
        if (!podman.isLocalOrTarExporter(inputs.outputs) &&
            (inputs.platforms.length == 0 || podman.satisfies(podmanVersion, '>=3.1.0'))) {
            args.push('--iidfile', await podman.getImageIDFile());
        }
        await (0, core_1.asyncForEach)(inputs.labels, async (label) => {
            args.push('--label', label);
        });
        await (0, core_1.asyncForEach)(inputs.outputs, async (output) => {
            args.push('--output', output);
        });
        if (inputs.platforms.length > 0) {
            args.push('--platform', inputs.platforms.join(','));
        }
        await (0, core_1.asyncForEach)(inputs.secrets, async (secret) => {
            try {
                args.push('--secret', await podman.getSecretString(secret));
            }
            catch (err) {
                core_1.logger.warn(this.getErrorMessage(err));
            }
        });
        await (0, core_1.asyncForEach)(inputs.secretFiles, async (secretFile) => {
            try {
                args.push('--secret', await podman.getSecretFile(secretFile));
            }
            catch (err) {
                core_1.logger.warn(this.getErrorMessage(err));
            }
        });
        if (inputs.githubToken && !podman.hasGitAuthToken(inputs.secrets) && context.startsWith(defaultContext)) {
            args.push('--secret', await podman.getSecretString(`GIT_AUTH_TOKEN=${inputs.githubToken}`));
        }
        if (inputs.shmSize) {
            args.push('--shm-size', inputs.shmSize);
        }
        await (0, core_1.asyncForEach)(inputs.ssh, async (ssh) => {
            args.push('--ssh', ssh);
        });
        await (0, core_1.asyncForEach)(inputs.tags, async (tag) => {
            args.push('--tag', tag);
        });
        if (inputs.target) {
            args.push('--target', inputs.target);
        }
        await (0, core_1.asyncForEach)(inputs.ulimit, async (ulimit) => {
            args.push('--ulimit', ulimit);
        });
        return args;
    }
    async getCommonArgs(inputs, _podmanVersion) {
        const args = [];
        if (inputs.load) {
            args.push('--load');
        }
        if (inputs.network) {
            args.push('--network', inputs.network);
        }
        if (inputs.noCache) {
            args.push('--no-cache');
        }
        if (inputs.pull) {
            args.push('--pull');
        }
        return args;
    }
}
exports.Podman = Podman;
