// @nx-tools/core/dist/lib/get-input-list.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getInputList = getInputList;
exports.getList = getList;
const sync_1 = require("csv-parse/sync");
const get_input_1 = require("./get-input");
function getInputList(name, opts) {
    return getList((0, get_input_1.getInput)(name, { prefix: opts?.prefix }), opts);
}
function getList(input, opts) {
    const res = [];
    if (input == '') {
        return opts?.fallback ?? res;
    }
    const records = (0, sync_1.parse)(input, {
        columns: false,
        relaxQuotes: true,
        comment: opts?.comment,
        relaxColumnCount: true,
        skipEmptyLines: true,
        quote: opts?.quote,
    });
    for (const record of records) {
        if (record.length == 1) {
            if (opts?.ignoreComma) {
                res.push(record[0]);
            }
            else {
                res.push(...record[0].split(','));
            }
        }
        else if (!opts?.ignoreComma) {
            res.push(...record);
        }
        else {
            res.push(record.join(','));
        }
    }
    return res.filter((item) => item).map((pat) => pat.trim());
}
