// @nx-tools/nx-container/dist/executors/build/context.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultContext = defaultContext;
exports.tmpDir = tmpDir;
exports.tmpNameSync = tmpNameSync;
exports.getInputs = getInputs;
exports.getInputList = getInputList;
exports.setOutput = setOutput;
const tslib_1 = require("tslib");
const core = tslib_1.__importStar(require("@nx-tools/core"));
const devkit_1 = require("@nx/devkit");
const sync_1 = require("csv-parse/sync");
const fs = tslib_1.__importStar(require("node:fs"));
const os = tslib_1.__importStar(require("node:os"));
const path = tslib_1.__importStar(require("node:path"));
const tmp = tslib_1.__importStar(require("tmp"));
let _defaultContext, _tmpDir;
function defaultContext() {
    if (!_defaultContext) {
        _defaultContext = '.';
    }
    return _defaultContext;
}
function tmpDir() {
    if (!_tmpDir) {
        _tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'docker-build-push-')).split(path.sep).join(path.posix.sep);
    }
    return _tmpDir;
}
function tmpNameSync(options) {
    return tmp.tmpNameSync(options);
}
async function getInputs(defaultContext, options, ctx) {
    const prefix = (0, devkit_1.names)(ctx?.projectName || '').constantName;
    return {
        quiet: core.getBooleanInput('quiet', { prefix, fallback: `${options.quiet || false}` }),
        addHosts: await getInputList('add-hosts', prefix, options['add-hosts']),
        allow: await getInputList('allow', prefix, options.allow),
        buildArgs: await getInputList('build-args', prefix, options['build-args'], true),
        buildContexts: await getInputList('build-contexts', prefix, options['build-contexts'], true),
        builder: core.getInput('builder', { prefix, fallback: options.builder }),
        cacheFrom: await getInputList('cache-from', prefix, [options['cache-from'] ?? []].flat(), true),
        cacheTo: await getInputList('cache-to', prefix, [options['cache-to'] ?? []].flat(), true),
        cgroupParent: core.getInput('cgroup-parent', { prefix, fallback: options['cgroup-parent'] }),
        context: core.getInput('context', { prefix, fallback: options.context || defaultContext }),
        file: core.getInput('file', { prefix, fallback: options.file }),
        githubToken: core.getInput('github-token'),
        labels: await getInputList('labels', prefix, options.labels, true),
        load: core.getBooleanInput('load', { fallback: `${options.load || false}` }),
        network: core.getInput('network', { prefix, fallback: options.network }),
        noCache: core.getBooleanInput('no-cache', { fallback: `${options['no-cache'] || false}` }),
        noCacheFilters: await getInputList('no-cache-filters', prefix, options['no-cache-filters']),
        outputs: await getInputList('outputs', prefix, options.outputs, true),
        platforms: await getInputList('platforms', prefix, options.platforms),
        provenance: core.getInput('provenance'),
        pull: core.getBooleanInput('pull', { fallback: `${options.pull || false}` }),
        push: core.getBooleanInput('push', { fallback: `${options.push || false}` }),
        sbom: core.getBooleanInput('sbom', { fallback: `${options.sbom || false}` }),
        secretFiles: await getInputList('secret-files', prefix, options['secret-files'], true),
        secrets: await getInputList('secrets', prefix, options.secrets, true),
        shmSize: core.getInput('shm-size', { prefix, fallback: options['shm-size'] }),
        ssh: await getInputList('ssh', prefix, options.ssh),
        tags: await getInputList('tags', prefix, options.tags),
        target: core.getInput('target', { prefix, fallback: options.target }),
        ulimit: await getInputList('ulimit', prefix, options.ulimit, true),
    };
}
async function getInputList(name, prefix, fallback, ignoreComma) {
    const res = [];
    const items = core.getInput(name, { prefix });
    if (items == '') {
        return fallback ?? res;
    }
    const records = await (0, sync_1.parse)(items, {
        columns: false,
        relaxQuotes: true,
        relaxColumnCount: true,
        // skipLinesWithEmptyValues: true,
        skipEmptyLines: true,
    });
    for (const record of records) {
        if (record.length == 1) {
            res.push(record[0]);
            continue;
        }
        else if (!ignoreComma) {
            res.push(...record);
            continue;
        }
        res.push(record.join(','));
    }
    return res.filter((item) => item).map((pat) => pat.trim());
}
function setOutput(name, value, ctx) {
    if (ctx?.projectName) {
        const outputDir = `${ctx?.root}/node_modules/.cache/nx-container/${ctx.projectName}`;
        fs.mkdirSync(outputDir, { recursive: true });
        fs.writeFileSync(`${outputDir}/${name}`, value);
    }
}
