// @nx-tools/nx-container/dist/generators/configuration/generator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.configurationGenerator = configurationGenerator;
exports.hasContainerPlugin = hasContainerPlugin;
const tslib_1 = require("tslib");
const devkit_1 = require("@nx/devkit");
const path = tslib_1.__importStar(require("node:path"));
const constants_1 = require("./constants");
function addFiles(tree, project, template) {
    const templateOptions = {
        projectName: project.name,
        template: '',
    };
    (0, devkit_1.generateFiles)(tree, path.join(__dirname, 'files', template), project.root, templateOptions);
}
async function configurationGenerator(tree, options) {
    const project = (0, devkit_1.readProjectConfiguration)(tree, options.project);
    if (!hasContainerPlugin(tree)) {
        (0, devkit_1.updateProjectConfiguration)(tree, options.project, {
            ...project,
            targets: {
                ...project.targets,
                container: {
                    executor: `@nx-tools/nx-container:build`,
                    dependsOn: ['build'],
                    options: {
                        engine: options.engine ?? constants_1.DEFAULT_ENGINE,
                        metadata: {
                            images: [project.name],
                            load: true,
                            tags: [
                                'type=schedule',
                                'type=ref,event=branch',
                                'type=ref,event=tag',
                                'type=ref,event=pr',
                                'type=sha,prefix=sha-',
                            ],
                        },
                    },
                },
            },
        });
    }
    addFiles(tree, project, options.template ?? constants_1.DEFAULT_TEMPLATE);
    if (!options.skipFormat) {
        await (0, devkit_1.formatFiles)(tree);
    }
}
function hasContainerPlugin(tree) {
    const nxJson = (0, devkit_1.readNxJson)(tree);
    return (nxJson?.plugins?.some((p) => typeof p === 'string' ? p === '@nx-tools/nx-container' : p.plugin === '@nx-tools/nx-container') ?? false);
}
exports.default = configurationGenerator;
