// @nx-tools/ci-context/dist/lib/utils/azure-devops.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Azure = void 0;
exports.repo = repo;
class Azure {
    static async context() {
        return {
            name: 'AZURE',
            actor: process.env['BUILD_SOURCEVERSIONAUTHOR'],
            eventName: process.env['SYSTEM_PULLREQUEST_PULLREQUESTID'] ? 'pull_request' : 'unknown',
            job: process.env['AGENT_JOBNAME'],
            payload: {},
            ref: process.env['SYSTEM_PULLREQUEST_SOURCEBRANCH'] || process.env['BUILD_SOURCEBRANCH'],
            runId: parseInt(process.env['BUILD_BUILDID'], 10),
            runNumber: parseInt(process.env['BUILD_BUILDID'], 10),
            repoUrl: process.env['BUILD_REPOSITORY_URI'],
            sha: process.env['BUILD_SOURCEVERSION'],
        };
    }
}
exports.Azure = Azure;
async function repo() {
    return {
        default_branch: '',
        description: '',
        html_url: process.env['BUILD_REPOSITORY_URI'],
        license: null,
        name: process.env['AGENT_JOBNAME'],
    };
}
