// @nx-tools/ci-context/dist/lib/utils/circle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Circle = void 0;
exports.repo = repo;
class Circle {
    static async context() {
        return {
            name: 'CIRCLE',
            actor: process.env['CIRCLE_USERNAME'],
            eventName: process.env['CI_PULL_REQUEST'] ? 'pull_request' : 'unknown',
            job: process.env['CIRCLE_JOB'],
            payload: {},
            ref: process.env['CIRCLE_TAG']
                ? `refs/tags/${process.env['CIRCLE_TAG']}`
                : `refs/heads/${process.env['CIRCLE_BRANCH']}`,
            runId: parseInt(process.env['CIRCLE_BUILD_NUM'], 10),
            runNumber: parseInt(process.env['CIRCLE_BUILD_NUM'], 10),
            repoUrl: process.env['CIRCLE_REPOSITORY_URL'],
            sha: process.env['CIRCLE_SHA1'],
        };
    }
}
exports.Circle = Circle;
async function repo() {
    return {
        default_branch: '',
        description: '',
        html_url: process.env['CIRCLE_REPOSITORY_URL'],
        license: null,
        name: process.env['CIRCLE_PROJECT_REPONAME'],
    };
}
