// @nx-tools/ci-context/dist/lib/utils/drone.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Drone = void 0;
exports.repo = repo;
class Drone {
    static async context() {
        return {
            name: 'DRONE',
            actor: process.env['DRONE_COMMIT_AUTHOR'],
            eventName: process.env['DRONE_BUILD_EVENT'],
            job: process.env['DRONE_STAGE_NAME'],
            payload: {
                repository: {
                    private: process.env['DRONE_REPO_PRIVATE'] === 'true',
                },
            },
            ref: process.env['DRONE_COMMIT_REF'],
            runId: parseInt(process.env['DRONE_BUILD_NUMBER'], 10),
            runNumber: parseInt(process.env['DRONE_BUILD_NUMBER'], 10),
            repoUrl: process.env['DRONE_REPO_LINK'],
            sha: process.env['DRONE_COMMIT_SHA'],
        };
    }
}
exports.Drone = Drone;
async function repo() {
    return {
        default_branch: process.env['DRONE_REPO_BRANCH'],
        description: '',
        html_url: process.env['DRONE_REPO_LINK'],
        license: null,
        name: process.env['DRONE_REPO'],
    };
}
