// @nx-extensions/helm/src/generators/dependency/generator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.dependencyGenerator = dependencyGenerator;
const tslib_1 = require("tslib");
const devkit_1 = require("@nx/devkit");
const js_yaml_1 = tslib_1.__importDefault(require("js-yaml"));
function dependencyGenerator(tree, options) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        var _a;
        const project = (0, devkit_1.readProjectConfiguration)(tree, options.project);
        if (!((_a = project.targets) === null || _a === void 0 ? void 0 : _a.helm)) {
            throw new Error(`Project ${options.project} does not have a helm target. Please run the chart generator first.`);
        }
        (0, devkit_1.updateProjectConfiguration)(tree, options.project, addDependencyToConfig(project, options.repositoryName, options.repository));
        updateChartYaml(tree, project, options.chartName, options.chartVersion, options.repository);
        if (options.format) {
            yield (0, devkit_1.formatFiles)(tree);
        }
    });
}
exports.default = dependencyGenerator;
function addDependencyToConfig(project, name, url) {
    return Object.assign(Object.assign({}, project), { targets: Object.assign(Object.assign({}, project.targets), { helm: Object.assign(Object.assign({}, project.targets.helm), { options: Object.assign(Object.assign({}, project.targets.helm.options), { dependencies: Object.assign(Object.assign({}, project.targets.helm.options.dependencies), { repositories: [
                            ...project.targets.helm.options.dependencies.repositories,
                            { name: name, url: url },
                        ] }) }) }) }) });
}
function updateChartYaml(tree, project, name, version, repository) {
    const chartFolder = project.targets.helm.options.chartFolder;
    const chartPath = `${chartFolder}/Chart.yaml`;
    if (!tree.exists(chartPath)) {
        throw new Error('Chart.yaml not found');
    }
    try {
        const chartContents = js_yaml_1.default.load(tree.read(chartPath, 'utf8').toString());
        if (!chartContents.dependencies) {
            chartContents.dependencies = [];
        }
        const existingDependency = chartContents.dependencies.find((dep) => dep.name === name);
        if (existingDependency) {
            existingDependency.version = version;
            existingDependency.repository = repository;
        }
        else {
            chartContents.dependencies.push({
                name: name,
                version: version,
                repository: repository,
            });
            tree.write(chartPath, js_yaml_1.default.dump(chartContents));
        }
    }
    catch (error) {
        console.error(error);
        throw new Error('Failed to parse Chart.yaml');
    }
}
