// @nx-tools/nx-container/dist/executors/build/engines/kaniko/kaniko.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getImageIDFile = getImageIDFile;
exports.getImageID = getImageID;
exports.getMetadataFile = getMetadataFile;
exports.getMetadata = getMetadata;
exports.getDigest = getDigest;
exports.isAvailable = isAvailable;
exports.getCommand = getCommand;
const tslib_1 = require("tslib");
const core = tslib_1.__importStar(require("@nx-tools/core"));
const node_fs_1 = tslib_1.__importDefault(require("node:fs"));
const node_path_1 = tslib_1.__importDefault(require("node:path"));
const context = tslib_1.__importStar(require("../../context"));
async function getImageIDFile() {
    return node_path_1.default.join(context.tmpDir(), 'iidfile').split(node_path_1.default.sep).join(node_path_1.default.posix.sep);
}
async function getImageID() {
    const iidFile = await getImageIDFile();
    if (!node_fs_1.default.existsSync(iidFile)) {
        return undefined;
    }
    return node_fs_1.default.readFileSync(iidFile, { encoding: 'utf-8' }).trim();
}
async function getMetadataFile() {
    return node_path_1.default.join(context.tmpDir(), 'metadata-file').split(node_path_1.default.sep).join(node_path_1.default.posix.sep);
}
async function getMetadata() {
    const metadataFile = await getMetadataFile();
    if (!node_fs_1.default.existsSync(metadataFile)) {
        return undefined;
    }
    const content = node_fs_1.default.readFileSync(metadataFile, { encoding: 'utf-8' }).trim();
    if (content === 'null') {
        return undefined;
    }
    return content;
}
async function getDigest(metadata) {
    if (metadata === undefined) {
        return undefined;
    }
    const metadataJSON = JSON.parse(metadata);
    if (metadataJSON['containerimage.digest']) {
        return metadataJSON['containerimage.digest'];
    }
    return undefined;
}
async function isAvailable() {
    const cmd = getCommand(['version']);
    return await core
        .getExecOutput(cmd.command, cmd.args, {
        ignoreReturnCode: true,
        silent: true,
    })
        .then((res) => {
        if (res.stderr.length > 0 && res.exitCode != 0) {
            return false;
        }
        return res.exitCode == 0;
    })
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        .catch((error) => {
        return false;
    });
}
function getCommand(args) {
    return {
        command: '/kaniko/executor',
        args,
    };
}
